processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -300,10 +323,172 @@ public int compare(Object lhs, Object rhs)
     );
   }
 
-  public Ordering<Row> getRowOrdering(final boolean granular)
+  private boolean validateAndGetForceLimitPushDown()
+  {
+    final boolean forcePushDown = getContextBoolean(GroupByQueryConfig.CTX_KEY_FORCE_LIMIT_PUSH_DOWN, false);
+    if (forcePushDown) {
+      if (!(limitSpec instanceof DefaultLimitSpec)) {
+        throw new IAE("When forcing limit push down, a limit spec must be provided.");
+      }
+
+      if (((DefaultLimitSpec) limitSpec).getLimit() == Integer.MAX_VALUE) {
+        throw new IAE("When forcing limit push down, the provided limit spec must have a limit.");
+      }
+
+      for (OrderByColumnSpec orderBySpec : ((DefaultLimitSpec) limitSpec).getColumns()) {
+        if (OrderByColumnSpec.getPostAggIndexForOrderBy(orderBySpec, postAggregatorSpecs) > -1) {
+          throw new UnsupportedOperationException("Limit push down when sorting by a post aggregator is not supported.");
+        }
+      }
+    }
+    return forcePushDown;
+  }
+
+  public boolean determineApplyLimitPushDown()
```

Comment:
i think this can be implemented in one line `return validateAndGetForceLimitPushDown() || DefaultLimitSpec.sortingOrderHasNonGroupingFields(..);`

you might have to add the  `defaultLimitSpec.getLimit() == Integer.MAX_VALUE` check inside `DefaultLimitSpec.sortingOrderHasNonGroupingFields(..)`